[EDITS]Copy to edit. (#1294)
mulesoft/docs-connectors: salesforce-composite/2.7/modules/ROOT/pages/edits/salesforce-composite-connector-examples.adoc

> = Salesforce Composite - Examples - Mule 4

Reviewer: ```suggestion
= Salesforce Composite Connector - Examples - Mule 4
```

> == SObject Tree Operation Example

Reviewer: You will want the narrative to this title to summarize the example and prepare the audience for EACH of the subtopics contained within. Take a look at all the === and evaluate whether they are introduced here.

> This example uses the following operations to connect to Salesforce, create and execute an SObject tree operation, and transform and output the results of the executed operation in JSON format:

Reviewer: "uses operations to create and execute an operation"?
```suggestion
This example uses the following operations to connect to Salesforce, create and execute an SObject tree, and transform and output the results in JSON format:
```

Reviewer: GLOBAL: If this is an operation name, it should use title-style capitalization, per style ref., not boldface.
Boldface is for GUI labels.

* HTTP *Listener*
+
Accepts data from HTTP requests
* *Transform Message*
+
Transforms the HTTP input

> * *Create sobject tree*

Reviewer: If this is an operation name, it should use title-style capitalization, per style ref.

+
Connects with Salesforce and creates an SObject tree
* *Transform Message*
+
Outputs the results of the *Create sobject tree* operation in JSON format
+
.The flow for this example includes an HTTP Listener component, Transform components, and a Create operation.
image::salesforce-composite-studio-use-case.png[Studio use case]

> 

Reviewer: So...this topic goes from listing goals to listing the operation names that achieve those goals to showing the flow.
I wonder if it might make more sense to introduce the goals, show the flow, and then define the operations used?

=== Add an HTTP Listener

> The HTTP *Listener* source operation initiates the flow when it detects an event in the `createSObjectTree` path:

Reviewer: This is a statement that tells what the operation does, but it does not mention "you," the audience, or why that audience performs the following steps. Clarify.

Reviewer: Typically, we introduce task steps with something that states directly or implies  "...to do <this>, follow these steps:".

. Create a new Mule project in Studio.

> . In the *Mule Palette* view, search for *HTTP* and select the *Listener* operation.

Reviewer: ```suggestion
. In the *Mule Palette* view, search for `http` and select the *Listener* operation.
```
The search is performed using lowercase letters (default: lettercase is irrelevant) that are user input ( ` ` ).
* * is reserved for GUI labels.

Reviewer: See Line 32 in the final file below.

. Drag the *Listener* operation onto the canvas.
. In the Listener properties window, set the *Path* field to `/createSObjectTree`.

=== Add the First Transform Message Component

> 

Reviewer: Does the audience understand why we are specifying "the first" and how many there are? Consider explaining this following the subtitle.

> This *Transform Message* component uses input data to transform the HTTP input for the Create operation:

Reviewer: Component names use sentence-style caps. but are not boldface unless referring specifically to the label in the UI.
```suggestion
The first Transform Message component uses input data to transform the HTTP input for the Create operation:
```

Reviewer: Also, is this clear: "uses input data to transform input for the operation"?

> . In the *Mule Palette* view, search for *Transform Message*.

Reviewer: ```suggestion
. In the *Mule Palette* view, search for `transform Message`.
```

. Drag the *Transform Message* component onto the canvas, to the right of *Listener*.
. Enter the name `Transform HTTP input for Create SObject Tree`.

> The following is an example of HTTP input for the *Transform Message* component:

Reviewer: ```suggestion
The following is an example of HTTP input for the Transform Message component:
```

         "attributes":{
            "objectType":"Account",
            "referenceId":"ref4"
         },
         "NumberOfEmployees":"101",
         "Name":"CreateSobjectTreeAccount2"
      },
      {
         "Contacts":{
            "records":[
               {
                  "Email":"sample@salesforce.com",
                  "Title":"President",
                  "attributes":{
                     "objectType":"Contact",
                     "referenceId":"ref2"
                  },
                  "LastName":"Smith"
               },
               {
                  "Email":"sample@salesforce.com",
                  "Title":"Vice President",
                  "attributes":{
                     "objectType":"Contact",
                     "referenceId":"ref3"
                  },
                  "LastName":"Evans"
               }
            ]
         },
         "Phone":"1234567890",
         "Website":"www.salesforce.com",
         "attributes":{
            "objectType":"Account",
            "referenceId":"ref1"
         },
         "NumberOfEmployees":"100",
         "Name":"CreateSobjectTreeAccount1"
      }
   ]
}
----

> 

Reviewer: BREAK HERE

Reviewer: RESUME edit

=== Add the Create SObject Tree Operation

> Add the *Create sobject tree* operation to connect with Salesforce and create an SObject tree:

Reviewer: ```suggestion
Add the Create SObject Tree operation to connect with Salesforce and create an SObject tree:
```
GLOBALLY evaluate: use bold and lowercase only when quoting the UI label.

> . In the *Mule Palette* view, search for *Salesforce* and select the *Create sobject tree* operation.

Reviewer: GLOBALLY reformat text input to search on to lowercase and monospace. And you can avoid some of the confusion by using select or click *this label*, without the "the" or object type:
```suggestion
. In the *Mule Palette* view, search for `salesforce` and select *Create sobject tree*.
```

Reviewer: See Line 32 in the final file below.

> . Drag the *Create sobject tree* operation to the right of *Transform Message*.

Reviewer: ```suggestion
. Drag *Create sobject tree* to the right of *Transform Message*.
```

> . Click the green plus sign (*+*) to the right of the *Connector configuration* field to access the Salesforce Composite global element configuration fields.

Reviewer: ```suggestion
. Click the green plus sign (*+*) to the right of *Connector configuration* to access the Salesforce composite global element configuration fields.
```

Reviewer: You can word steps either way, and the Connectors template typically uses "the <label> <element type>," but, as I said, it might be easier to omit the type of element in favor of the label alone.

. Complete the fields using variables, as follows:
.. In the *Create sobject tree* properties window, enter `Account` in the *SObject Root Type* field.

> .. In the properties file `src/main/mule/mule-app.properties`, declare and enter values for the variables in the global element.

Reviewer: Is there a reason to have this step with two substeps rather than two primary steps?

=== Add the Second Transform Message Component

> This *Transform Message* component transforms the output result into a JSON:

Reviewer: ```suggestion
The second Transform Message component transforms the output result into JSON:
```
JSON (JavaScript Object Notation) is a programming language, not a thing. Unless you mean a JSON file?

> . Drag a second *Transform Message* component to the right of *Create sobject tree*.

Reviewer: Perhaps...:
```suggestion
. Drag *Transform Message* again to the right of *Create sobject tree*.
```

. Enter the following:
+
[source,dataweave,linenums]
----
%dw 2.0
%output application/json
payload
---
{
"Id": payload.Id,
"Fields" : ["Id", "Name"],
"Type": "Account"
}
----

=== XML for the SObject Tree Operation Flow

> Paste this XML code in to the *Configuration XML* tab of the Anypoint Studio canvas to experiment with the flow:

Reviewer: into
```suggestion
Paste this XML code into the *Configuration XML* tab of the Anypoint Studio canvas to experiment with the flow:
```

	"NumberOfEmployees": record.NumberOfEmployees as Number,
	("ChildAccounts": {
		"records": record.ChildAccounts.records map ((record01, indexOfRecord01) -> {
			"attributes": {
				"type": record01.attributes.objectType,
				"referenceId": record01.attributes.referenceId
			},
			"Name": record01.Name,
			"Phone": record01.Phone,
			"Website": record01.Website,
			"NumberOfEmployees": record01.NumberOfEmployees as Number
		})
	}),
	"Contacts": {
		"records": record.Contacts.records map ((record01, indexOfRecord01) -> {
			"attributes": {
				"type": record01.attributes.objectType,
				"referenceId": record01.attributes.referenceId
			},
			"LastName": record01.LastName,
			"Email": record01.Email,
			"Title": record01.Title
		})
	}
})
]]></ee:set-payload>
			</ee:message>
		</ee:transform>
		<salesforce-composite:create-sobject-tree
      doc:name="Create SObject tree"
      config-ref="Salesforce_Composite_Salesforce_Composite"
      type="Account"/>
		<ee:transform doc:name="Transform Message">
			<ee:message >
				<ee:set-payload ><![CDATA[%dw 2.0
output application/json
---
payload]]></ee:set-payload>
			</ee:message>
		</ee:transform>
	</flow>
</mule>
----

> == Pre-Query Example

Reviewer: pre-query or pre-create?
When use as the operation name, Pre-create or Pre-Create?
GLOBAL

This example provides metadata for an object to query, generates two subrequests that contain this metadata, and then executes the subrequests:

* HTTP *Listener*
+
Accepts data from HTTP requests
* *Transform Message*
+

> Specifies name for object

Reviewer: ```suggestion
Specifies a name for the object
```
Avoid "shorthand" language

* *Pre create*
+

> Provides metadata for the object that is going to be created and generates a subrequest used in execution

Reviewer: Here and below, I think "the object that is going to be" can be "the object to be"

* *Transform Message*
+
Specifies values for fields
* *Pre query*
+
Provides metadata for the object that is going to be queried and generates a subrequest used in execution
* *Transform Message*
+
Passes output from *Pre query* operation
* *Execute composite batch*
+
Sends all subrequests in one batch
* *Transform Message*
+
Outputs the results in JSON format
+
.The flow for this example includes an HTTP Listener component, Transform components, and multiple Salesfore Composite operations.
image::salesforce-composite-prequery-flow.png[Pre-Query flow in Studio]

=== Add an HTTP Listener

> The HTTP *Listener* source operation initiates the flow when it detects an event in the `executeFlow` path:

Reviewer: This tells what the operation does and ends with a colon indicating that the following steps tell what the operation does, not what the audience does. Reword for clarity.

. Create a new Mule project in Studio.

> . In the *Mule Palette* view, search for *HTTP* and select the *Listener* operation.

Reviewer: ```suggestion
. In the *Mule Palette* view, search for `http` and select the *Listener* operation.
```

. Drag the *Listener* operation onto the canvas.
. In the Listener properties window, set the *Path* field to `/executeFlow`.

=== Add the First Transform Message Component

> This *Transform Message* component specifies a name for the created object and passes the name to the *Pre create* operation:

Reviewer: ```suggestion
The first Transform Message component specifies a name for the created object and passes the name to the Pre-create operation:
```

> . In the *Mule Palette* view, search for *Transform Message*.

Reviewer: ```suggestion
. In the *Mule Palette* view, search for `transform message`.
```

> . Drag the *Transform Message* component onto the canvas, to the right of *Listener*.

Reviewer: ```suggestion
. Drag *Transform Message* onto the canvas, to the right of *Listener*.
```

. Enter the name `MyNewAccount`:
+
[source,dataweave,linenums]
----
%dw 2.0
%output application/java
---
{
   Name: "MyNewAccount"
}
----

> === Add the Pre Create Operation

Reviewer: ```suggestion
=== Add the Pre-Create Operation
```

> The *Pre create* operation provides metadata for creating an object called `NewAccount` and generates a subrequest for the `executeCompositeBatch` operation:

Reviewer: ```suggestion
The Pre-Create operation provides metadata for creating an object called `NewAccount` and generates a subrequest for the `executeCompositeBatch` operation:
```

> . In the *Mule Palette* view, search for *Salesforce* and select the Salesforce Composite *Pre create* operation.

Reviewer: GLOBALLY fix the "search for" strings to lowercase mono; I'll stop marking.

Reviewer: Also GLOBALLY fix the names of the operations (plain init. cap.) vs. the UI labels for the UI elements (bold). I'll stop marking.

. Drag the *Pre create* operation to the right of *Transform Message*.
. Click the green plus icon (*+*) to the right of the *Connector configuration* field to access the Salesforce Composite global element configuration fields.
. Complete the fields using variables, as follows:
.. In the *Pre create* properties window, enter `Account` in the *Type* field.

> .. In the properties file `src/main/mule/mule-app.properties`, declare and enter values for the variables in the global element.

Reviewer: ditto question above, Lines 126-8

+
.Complete fields in the General section to create a reusable global element for Salesforce Composite Connector.
image::salesforce-composite-global-config.png[Salesforce Composite Connector global element configuration]
+

=== Add the Second Transform Message Component

> This *Transform Message* component specifies values for the *ID* and *Type* fields, and then passes these values to the *Pre query* operation:

Reviewer: I suggest replacing "this" with the specific count: the first, the second, the third. This helps locate the user in the flow.
GLOBAL: I have not marked each instance.

. Drag a second *Transform Message* component to the right of *Pre create*.
. Enter the following name-value pairs:
+
[source,dataweave,linenums]
----
%dw 2.0
%output application/java
---
{
   "Id": payload.Id,
   "Fields" : ["Id", "Name"],
   "Type": "Account"
}
----

=== Add the Pre Query Operation

The *Pre query* operation provides metadata for an object to be queried. It generates a subrequest for the *Execute composite batch* operation:

. Drag a Salesforce Composite *Pre query* operation to the right of the second *Transform Message*.
. In the *Query* field, enter the following:
+
`Select Name from Account WHERE Name LIKE '%:name %'`

=== Add the Third Transform Message Component

This *Transform Message* operation passes the output of the *Pre query* operation to the *Execute composite batch* operation:

. Drag a third *Transform Message* component to the right of *Pre query*.
. Set the output to `payload`:
+
[source,dataweave,linenums]
----
%dw 2.0
%output application/java
---
[
payload
]
----

=== Add the Execute Composite Batch Operation

> The *Execute composite batch* operation executes the subrequests that the *Pre create* and *Pre query* operations created:

Reviewer: Ohhhh...two pre things!

---
{
   "Id": payload.Id,
   "Fields" : ["Id", "Name"],
   "Type": "Account"
}]]></ee:set-payload>
         </ee:message>
      </ee:transform>
      <salesforce-composite:pre-query doc:name="Pre query" config-ref="Salesforce_Composite_Config">
         <salesforce-composite:query >Select Name from Account WHERE Name LIKE '%:name %'
                    </salesforce-composite:query>
         <salesforce-composite:parameters ><![CDATA[#[output application/java
---
{
   "name" : "MyNewAccount"
}]]]></salesforce-composite:parameters>
      </salesforce-composite:pre-query>
      <ee:transform doc:name="Transform Message" >
         <ee:message >
            <ee:set-payload ><![CDATA[%dw 2.0
output application/java
---
[
   payload
]]]></ee:set-payload>
         </ee:message>
      </ee:transform>
      <salesforce-composite:execute-composite-batch doc:name="Execute composite batch" config-ref="Salesforce_Composite_Config"/>
      <ee:transform doc:name="Transform Message" >
         <ee:message >
            <ee:set-payload ><![CDATA[%dw 2.0
output application/json
---
payload]]></ee:set-payload>
         </ee:message>
      </ee:transform>
   </flow>
</mule>
----

== See Also

> * xref:salesforce-composite-connector-reference.adoc[Salesforce Composite Connector Reference]

Reviewer: Are all previous uses of Salesforce Composite referring to the connector? Clarify.

> * https://help.mulesoft.com[MuleSoft Help Center]

Reviewer: BREAK and commit
